Sources/Transformer.swift

```diff
@@ -0,0 +1,177 @@
+//  of this software and associated documentation files (the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions:
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import Foundation
+import AlgoliaSearch
+
+/// A Transformable data source
+public protocol Transformable {
+    associatedtype IResults
+    associatedtype IParameters
+    
+    typealias SearchResultsHandler = (_ results: IResults?, _ error: Error?) -> Void
+    
+    // Search operation
+    func search(_ query: IParameters, searchResultsHandler: @escaping SearchResultsHandler)
+    
+    // Transforms the Algolia params to custom backend params.
+    func map(query: Query) -> IParameters
+    
+    // Transforms the custom backend result to an Algolia result.
+    func map(results: IResults) -> SearchResults
+    
+    
+    // Disjunctive Search operation
+    func searchDisjunctiveFaceting(_ query: IParameters, searchResultsHandler: @escaping SearchResultsHandler)
+    
+    // Transforms the Algolia disjunctive params to custom backend params.
+    func map(query: Query, disjunctiveFacets: [String], refinements: [String : [String]]) -> IParameters
+    
+    
+    // Search for facet value operation
+    func searchForFacetValues(_ query: IParameters, searchResultsHandler: @escaping SearchResultsHandler)
+    
+    // Transforms the Algolia facet value params to custom backend params.
+    func map(query: Query?, facetName: String, matching text: String) -> IParameters
+}
+
+/// Base class used to implement a custom backend with Algolia
+/// You need to make sure to implement the following method:
+/// - search(query:searchResultsHandler:)
+/// - map(query:)
+/// - map(results:)
+open class SearchTransformer<Parameters, Results>: Transformable, Searchable {
+    
+    public typealias IParameters = Parameters
+    public typealias IResults = Results
+    
+    public init() {}
+    
+    // Transformable protocol
+    
+    open func search(_ query: Parameters, searchResultsHandler: @escaping SearchResultsHandler) {
+        fatalError("make sure to override search(query:searchResultsHandler:) for custom backend")
```

Comment:
This `fatalError` is only raised if the `search()` method is not implemented, right?

Reply:
Exactly: when someone tries to implement the `SearchTransformer` class, if he forget this method, he'll automatically get a crash. This is to force him to implement the method. The reason why it's done this way (which is a bit dirty) is due to Swift limitations that I described in the PR description

Reply:
Just wanted to make sure! I decided to return empty results (fail silently) but to warn in the console. Happy to go either way though.

I think this is fine to do like this in Swift.

Reply:
On second thought, you're right. I now throw errors if the method is not implemented.